src/rdkafka_transport.c

```diff
@@ -498,6 +500,14 @@ static int rd_kafka_transport_ssl_connect (rd_kafka_broker_t *rkb,
 	if (!SSL_set_fd(rktrans->rktrans_ssl, rktrans->rktrans_s))
 		goto fail;
 
+	mtx_lock(&rkb->rkb_lock);
+	rd_snprintf(name, sizeof(name), "%s", rkb->rkb_nodename);
+	mtx_unlock(&rkb->rkb_lock);
+	if ((t = strrchr(name, ':')))
+		*t = '\0';
+	if (!SSL_set_tlsext_host_name(rktrans->rktrans_ssl, name))
```

Comment:
You had me at "the official Java client".

Re hostname check, something like this should be enough:
```
if (!(/*ipv6*/((strchr(hostname, ':') && strspn(hostname, "0123456789abcdefABCDEF:.[]%") == strlen(hostname)) ||
    /*ipv4*/strspn(hostname, "01234567890.") == strlen(hostname)))
  is a hostname;
```